src/components/Personalization/index.js

```diff
@@ -12,43 +12,83 @@ governing permissions and limitations under the License.
 
 import { isNonEmptyArray } from "../../utils";
 import { initRuleComponentModules, executeRules } from "./turbine";
+import { hideContainers, showContainers, hideElements } from "./flicker";
 
 const PAGE_HANDLE = "personalization:page";
+const EVENT_COMMAND = "event";
+const isElementExists = event => event.moduleType === "elementExists";
 
-const createPersonalization = ({ logger }) => {
+const hideElementsForPage = fragments => {
+  fragments.forEach(fragment => {
+    const { rules = [] } = fragment;
+
+    rules.forEach(rule => {
+      const { events = [] } = rule;
+      const filteredEvents = events.filter(isElementExists);
+
+      filteredEvents.forEach(event => {
+        const { settings = {} } = event;
+        const { prehidingSelector } = settings;
+
+        if (prehidingSelector) {
+          hideElements(prehidingSelector);
+        }
+      });
+    });
+  });
+};
+
+const executeFragments = (fragments, modules, logger) => {
+  fragments.forEach(fragment => {
+    const { rules = [] } = fragment;
+
+    if (isNonEmptyArray(rules)) {
+      executeRules(rules, modules, logger);
+    }
+  });
+};
+
+const createPersonalization = ({ config, logger }) => {
+  const { prehidingStyle } = config;
   let ruleComponentModules;
 
   return {
     lifecycle: {
       onComponentsRegistered(tools) {
         const { componentRegistry } = tools;
         ruleComponentModules = initRuleComponentModules(
-          componentRegistry.getCommand("event")
+          componentRegistry.getCommand(EVENT_COMMAND)
         );
       },
       onBeforeEvent(event, isViewStart) {
         if (!isViewStart) {
           return;
         }
 
+        // For viewStart we try to hide the personalization containers
+        hideContainers(prehidingStyle);
+
         event.mergeQuery({
           personalization: {
             page: true,
             views: true
           }
         });
+
         event.expectResponse();
       },
       onResponse(response) {
         const fragments = response.getPayloadByType(PAGE_HANDLE) || [];
 
-        fragments.forEach(fragment => {
-          const { rules = [] } = fragment;
+        // On response we first hide all the elements for
+        // personalization:page handle
+        hideElementsForPage(fragments);
```

Comment:
Do we need to think about what happens if the request fails, or times out?

If so, do you think we need an `onError` lifecycle hooks so you can decide what to do in case of an error?

Reply:
@jfkhoury good call, do we have an `onResponseError` hook? That would help to ensure we un-hide the containers in case of network request failure.

Reply:
@artur-ciocanu, we currently have this TODO in the lifecycle module: `// TODO: We might need an `onError(error)` hook.`. We were waiting for the first use case before we implement it; and this seems like the perfect candidate ;)

Reply:
You can use the onBeforeDataCollection event for this.
```suggestion
const onBeforeDataCollection = (payload, promise) => {
  promise.catch(() => {
    ...
  });
};
```

But it's not easy to know which request it is for.

Reply:
@jfkhoury and @Aaronius as discussed I'll add a separate `onResponseError` so all this cleanup can be handled properly.